include/libhal-soft/i2c_bit_bang.hpp

```diff
@@ -0,0 +1,210 @@
+// limitations under the License.
+
+#pragma once
+
+#include <chrono>
+#include <libhal/i2c.hpp>
+#include <libhal/output_pin.hpp>
+#include <libhal/units.hpp>
+
+namespace hal {
+/**
+ * @brief A bit bang implementation for i2c. This implementation of i2c only
+ * needs 2 gpio's to work correctly. In its current implementation we only
+ * support single controller multiple peripheral. But have intentions of
+ * supporting multiple controller soon. Additionally, it's important to note
+ * that bit bang is a best try implementation of i2c. As in it will do it's best
+ * to reach the requested frequency but will typically be slower then requested
+ * with brief testing it seems that the error is about 2us per period for the
+ * lpc4078 , so you can try to account for this error with the frequency you set
+ * for example, achieving the correct period for 100kHz would result in setting
+ * slightly more then a 125kHz frequency. Additionally, the fastest frequency
+ * this bit bang can handle is around 200kHz.
+ */
+class i2c_bit_bang : public i2c
+{
+public:
+  /// holds all of the information for an i2c bus
+  struct bus_info
+  {
+    float duty_cycle = 0.5f;
+  };
+
+  struct i2c_pins
+  {
+    output_pin* sda;
+    output_pin* scl;
+  };
+
+  /**
+   * @brief Construct a new i2c bit bang object
+   *
+   * @param p_pins this holds both scl and sda to be used inside of the driver
+   * @param p_steady_clock a steady clock that should have a higher frequency
+   * then the configured frequency for the bit bang
+   * @param p_bus the bus information which will be used which configuring the
+   * bit bang interface
+   */
+
+  i2c_bit_bang(i2c_pins p_pins, steady_clock& p_steady_clock, bus_info p_bus);
+
+  void driver_configure(const settings& p_settings) override;
+
+  virtual void driver_transaction(
+    hal::byte p_address,
+    std::span<const hal::byte> p_data_out,
+    std::span<hal::byte> p_data_in,
+    function_ref<hal::timeout_function> p_timeout) override;
+
+private:
+  /**
+   * @brief this function will send the start condition and it will always pull
+   * the pins high before sending the start condition
+   */
+  void send_start_condition();
+
+  /**
+   * @brief this function will send the stop condition while also making sure
+   * the sda pin is pulled low before sending the stop condition
+   *
+   */
+  void send_stop_condition();
+
+  /**
+   * @brief this function will go through the steps of writing the address of
+   * the peripheral the controller wishes to speak to while also ensuring the
+   * data written is acknowledged
+   *
+   * @param p_address the address of the peripheral, configured with the
+   * read/write bit already that the controller is requesting to talk to
+   * @param p_timeout a timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   *
+   * @throws hal::io_error when the address written to the bus was not
+   * acknowledged
```

Comment:
```suggestion
   * @throws hal::no_such_device when the address written to the bus was not
   * acknowledged
```

It is not an io_error for a device to now respond on the bus. It just indicates that the device isn't present or responsive on the bus.

Reply:
Oops, must have been referencing the wrong function, thanks.